pam/internal/gdm/conversation_test.go

```diff
@@ -99,3 +129,569 @@ func TestSendToGdm(t *testing.T) {
+						require.NoError(t, err)
+						return pam.BinaryPointer(msg), nil
+					}
+					msg, err := newJSONProtoMessage(req)
+					require.NoError(t, err)
+					return pam.BinaryPointer(msg), tc.wantError
+				}))
+
+			data, err := tc.value.SendParsed(mtx)
+			require.True(t, convFuncCalled)
+
+			if tc.wantError != nil {
+				require.Nil(t, data)
+				require.Error(t, tc.wantError, err)
+				return
+			}
+
+			require.NoError(t, err)
+
+			if tc.wantReturn != nil {
+				require.Equal(t, tc.wantReturn, data)
+			} else {
+				require.Equal(t, tc.value, data)
+			}
+		})
+	}
+}
+
+func TestDataSendPoll(t *testing.T) {
+	t.Parallel()
+	t.Cleanup(pam_test.MaybeDoLeakCheck)
+
+	testCases := map[string]struct {
+		wantReturn *Data
+		wantError  error
+	}{
+		"empty return": {
+			wantReturn: &Data{
+				Type: DataType_pollResponse,
+			},
+		},
+
+		"return values": {
+			wantReturn: &Data{
+				Type: DataType_pollResponse,
+				PollResponse: []*EventData{
+					{Type: EventType_authEvent, Data: &EventData_AuthEvent{}},
+					{Type: EventType_authModeSelected, Data: &EventData_AuthModeSelected{}},
+					{Type: EventType_uiLayoutReceived, Data: &EventData_UiLayoutReceived{}},
+				},
+			},
+		},
+
+		"Error on nil return": {
+			wantReturn: nil,
+			wantError:  errors.New("unexpected token null"),
+		},
+
+		"Error on unexpected type": {
+			wantReturn: &Data{Type: DataType_hello},
+			wantError:  errors.New("gdm replied with an unexpected type: hello"),
+		},
+	}
+
+	for name, tc := range testCases {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			t.Cleanup(pam_test.MaybeDoLeakCheck)
+			convFuncCalled := true
+			mtx := pam_test.NewModuleTransactionDummy(pam.BinaryPointerConversationFunc(
+				func(ptr pam.BinaryPointer) (pam.BinaryPointer, error) {
+					convFuncCalled = true
+					require.NotNil(t, ptr)
+					if tc.wantReturn != nil {
+						bytes, err := tc.wantReturn.JSON()
+						require.NoError(t, err)
+						msg, err := newJSONProtoMessage(bytes)
+						require.NoError(t, err)
+						return pam.BinaryPointer(msg), nil
+					}
+					msg, err := newJSONProtoMessage([]byte("null"))
+					require.NoError(t, err)
+					return pam.BinaryPointer(msg), tc.wantError
+				}))
+
+			eventData, err := SendPoll(mtx)
+			require.True(t, convFuncCalled)
+
+			if tc.wantError != nil {
+				require.Nil(t, eventData)
+				require.Error(t, tc.wantError, err)
+				return
+			}
+
+			require.NoError(t, err)
+			requireEqualData(t, tc.wantReturn,
+				&Data{Type: DataType_pollResponse, PollResponse: eventData})
+		})
+	}
+}
+
+type invalidRequest struct {
+}
+
+// This is to implement Request interface defined by protobuf.
+//
+//nolint:revive
+func (*invalidRequest) isRequestData_Data() {}
+
+type sendRequestTypedCase struct {
+	request   Request
+	wantData  *Data
+	wantError error
+}
+
+func testDataSendRequestTyped[ResponseType Response](t *testing.T, specificCases map[string]sendRequestTypedCase) {
```

Comment:
Don't use `test` as a prefix for helper functions. This is either a test or a helper.

Reply:
Well, I did some other way (see commit f0a2617) and this avoids creating further functions, but it also duplicates code quite a bit... I'm not aware of any other way we can use when we want to test generics without doing manual repetitions that generics are supposed to avoid.